# Lista 1

## Uczenie maszynowe i sztuczna inteligencja

* [regresja](https://pl.wikipedia.org/wiki/Regresja_(statystyka))
* [podobieństwo konsinusowe](https://en.wikipedia.org/wiki/Cosine_similarity)


## Zadanie 1 (10pt)

Napisz wykorzystując regresję liniową program, który na podstawie oceny filmów przez użytkowników będzie
starał się przewidzieć ocenę innych użytkowników. Jako dane wykorzystamy zbiór
[MovieLens Latest Datasets](https://grouplens.org/datasets/movielens/latest/).
Dokładnie wybierzemy mniejszy zbiór, pobierz plik
[ml-latest-small.zip](http://files.grouplens.org/datasets/movielens/ml-latest-small.zip").
Zadanie polega na wybraniu z pliku <em>ratings.csv</em> tych użytkowników ($\mbox{userId}$), którzy ocenili film
'Toy Story (1995)', który w tym pliku ma identyfikator '1' (patrz <em>movies.csv</em>).
W tym pliku osób takich jest $215$. Wtedy zgodnie z zapisem z wykładu
$x_{ij}$ będzie oceną $i$-tego użytkownika dla $i=0,\ldots, 214$, bo
taki jest nasz zbiór osób, które oceniły 'Toy Story' oraz $j$ będzie oceną $j$-tego filmu
dla $j=0,\ldots,m$. Jako $j$ można wybrać $\mbox{movieId}$ filmu, czyli
np. film o $\mbox{movieId}=42$ oceniony przez użytkownika $5$ (nie jest to $\mbox{userId}$, tylko
piąta osoba ze zbioru $215$ osób), który ocenił film jako np. $3.5$ wpisujemy $x[5, 42] = 3.5$.
Natomiast $y_i$ to ocena 'Toy Story' przez $i$-tego użytkownika. Zatem tworzymy macierz $X = [x_{ij}]$ oraz wektor $y_i$
gdzie $i=0,\ldots,215$ oraz $j=0,\ldots,m$. Dla tak przygotowanych danych wykonujemy:

* regresje liniową na całym zbiorze użytkowników dla $m=10,1000,10000$,
  czyli np. dla $m=10$ ignorujemy filmy o $\mbox{movieId} \gt 10$ i robimy regresje
  dla tak okrojonego zbioru ocen. Jakie dostajemy błędy. Pokaż na wykresie.
* podziel zbiór osób na tzw. zbiór treningowy oraz zbiór testowy np.
  weźmy $i=0,\ldots,200$ to będzie zbiór treningowy i na takim zbiorze osób wykonajmy regresje
  natomiast później sprawdzamy już dla całości ($215$). Zatem ostatnie
  $15$ ocen będziemy chcieli przewidzieć (zbiór testowy). Zrób przewidywanie dla
  $m=10,100,200,500,1000,10000$. Wyświetl wynik predykcji i wynik prawidłowy
  dla tych $15$ osób.

Przykład danych do regresji jakie otrzymujemy dla $m=10$ i $n=215$.

```
X = [[0.  4.  0.  0.  4.  0.  0.  0.  0.  0. ]
     [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
     [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
     [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
     [0.  0.  0.  0.  0.  0.  0.  0.  0.  0. ]
     [3.  0.  0.  0.  4.  0.  0.  0.  0.  0. ]
     [3.  3.  0.  0.  0.  2.  0.  0.  2.  0. ]
     [3.5 0.  0.  0.  0.  0.  0.  0.  5.  0. ]
     ...
y = [[4. ]
     [4. ]
     [4.5]
     [2.5]
     [4.5]
     [3.5]
     [4. ]
     [3.5]
     ...
```


## Zadanie 2 (15pt)

Napisz system rekomendacji filmów. Systemy takie są wykorzystywane przez różne firmy np.
Netflix organizował konkurs na opracowanie algorytmu, który będzie przewidywał
ocenę użytkownika [Netflix Prize](https://en.wikipedia.org/wiki/Netflix_Prize).
W zadaniu tym zaimplementujemy podobny system, który jednak zamiast przewidywania będzie na
podstawie preferencji użytkownika rekomendował filmy, które najprawdopodobniej mu się spodobają.
Istnieje wiele sposobów, aby taki system napisać, dla zainteresowanych bardziej tematem proponuje
zobaczyć np. [Recommendation Systems](http://infolab.stanford.edu/~ullman/mmds/ch9.pdf).
W tym zadaniu wybierzemy w miarę prosty i łatwy do implementacji system rekomendacji.
Sformalizujmy problem. Załóżmy, że mamy macierz oceny, gdzie wiersze będą reprezentować
użytkowników a kolumny filmy np.
$$ \begin{array}{ccc}
    & \mbox{Matrix} & \mbox{Star Wars IV} \\
    \mbox{Alice} & 5 & 4 \\
    \mbox{Bob} & 0 & 1 \\
    \mbox{John} & 2 & 2 \\
    \mbox{Ada} & 5 & 5
   \end{array}$$
Patrząc na powyższą macierz widać, że kolumny pierwsza i druga mają podobne oceny stąd można wywnioskować, że
filmy Matrix i Star Wars IV są (według użytkowników) podobne do siebie, czyli jeśli komuś podobał się Matrix
to jest duża szansa, że będzie podobał mu się Star Wars i odwrotnie. Dlatego "podobieństwo" sformalizujemy przez wykorzystanie
[podobieństwa cosinusowego](https://en.wikipedia.org/wiki/Cosine_similarity), czyli
jeśli $x = (x_1, x_2, \dots, x_n)$ i $y = (y_1, y_2, \dots, y_n)$, wtedy:
$$x_1 y_1 + x_2 y_2 + \cdots + x_n y_n = x \cdot y = \lVert x \rVert \lVert y \rVert \cos(\theta)$$
Gdzie $\theta$ jest kątem między wektorami. Wygodniej będzie nam normalizować wektory. Wtedy
$$x_1 y_1 + x_2 y_2 + \cdots + x_n y_n = \cos(\theta).$$ Wtedy $\cos(\theta)$ reprezentuje
podobieństwo jednego wektora do drugiego, czyli jak bardzo jedne oceny są bliskie innym.

Dla naszej macierzy otrzymujemy

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
 
 ####################################### part 1 #######################################
# Wczytanie danych z pliku ratings.csv
ratings_df = pd.read_csv("ratings.csv")

# Wybieramy oceny użytkowników dla filmu 'Toy Story (1995)' (movieId=1)
toy_story_ratings = ratings_df[ratings_df['movieId'] == 1]

# Lista identyfikatorów użytkowników, którzy ocenili 'Toy Story'
toy_story_user_ids = toy_story_ratings['userId'].tolist()
#print(len(toy_story_user_ids))
#print(toy_story_user_ids)

# Wybieramy tylko te wiersze z ocenami, które odpowiadają użytkownikom oceniającym 'Toy Story'
selected_ratings = ratings_df[ratings_df['userId'].isin(toy_story_user_ids)]

# Tworzymy macierz ocen (X) i wektor ocen 'Toy Story' (y)
X = selected_ratings.pivot(index='userId', columns='movieId', values='rating').fillna(0)

X = X.drop(columns=[1])

y = selected_ratings[selected_ratings['movieId'] == 1].set_index('userId')['rating']
print(X.head(10))
#print(y.head(10))

# Lista różnych wartości m (liczba kolumn)
m_values = [10, 100, 1000, 10000]

# Lista, do której będziemy zapisywać błędy dla różnych wartości m
errors = []

# caly zbior z ocenami
X_all = ratings_df.pivot(index='userId', columns='movieId', values='rating').fillna(0)
X_all = X_all.drop(columns=[1])

"""
Wykonujemy regresje 
"""

# Dla każdej wartości m wykonujemy regresję liniową i obliczamy błąd
# Lista, do której będziemy zapisywać wartości współczynnika determinacji (R^2) dla różnych wartości m
r_squared_values = []
for m in m_values:
    # Wybieramy m pierwszych filmów
    first_movies = X.columns[:m]
    # Tworzymy X zawierające tylko m pierwszych filmów
    X_subset = X[first_movies]
    
    X_all_subset = X_all[first_movies]
    # Tworzymy model regresji liniowej
    model = LinearRegression()
    # Dopasowujemy model do danych, trenujemy model
    model.fit(X_subset, y)
    # Przewidujemy oceny 'Toy Story' dla WSZYSTKICH użytkowników
    y_pred = model.predict(X_all_subset)
    #print(len(y_pred))
    # ze zbioru predycji wybieramy tylko uzytkownikow ktorzy oceniali toy story
    y_pred_selected = y_pred[X_all_subset.index.isin(toy_story_user_ids)]
    
    #print("y_pred_selected: ", y_pred_selected)
    
    # Obliczamy błąd (MSE) dla przewidywanych ocen
    error = mean_squared_error(y, y_pred_selected)
    errors.append(error)
    #print("Errors: ",errors)
    
    # Obliczamy współczynnik determinacji (R^2) dla WSZYSTKICH użytkowników
    r_squared = r2_score(y, y_pred_selected)
    r_squared_values.append(r_squared)

# Wykres zależności błędu od liczby kolumn (wartości m)
plt.plot(m_values, errors, marker='o')
plt.title('Zależność błędu od liczby kolumn (wartości m) dla zbioru treningowego')
plt.xlabel('Liczba kolumn (m)')
plt.ylabel('Błąd (MSE)')
plt.grid(True)
plt.show()

plt.plot(m_values, r_squared_values, marker='o')
plt.title('Zależność współczynnika determinacji od liczby kolumn (wartości m) dla zbioru treningowego')
plt.xlabel('Liczba kolumn (m)')
plt.ylabel('Współczynnik determinacji (R^2)')
plt.grid(True)
plt.xscale('log')  # Ustawienie skali logarytmicznej na osi poziomej
plt.show()

####################################### part 2 #######################################

# Podział zbioru osób na zbiór treningowy i zbiór testowy
train_size = 200
X_train, X_test = X.iloc[:train_size], X.iloc[train_size:]
y_train, y_test = y.iloc[:train_size], y.iloc[train_size:]

# Lista różnych wartości m (liczba kolumn)
m_values = [10, 100, 200, 500, 1000, 10000]

# Lista, do której będziemy zapisywać predykcje i błędy dla różnych wartości m
predictions = {}

# Dla każdej wartości m wykonujemy regresję liniową i obliczamy predykcje oraz błąd na zbiorze testowym
for m in m_values:
    # Wybieramy m pierwszych filmów
    first_movies1 = X_train.columns[:m]
    # Tworzymy X zawierające tylko m pierwszych filmów
    X_train_subset = X_train[first_movies1]
    X_test_subset = X_test[first_movies1]
    
    # Tworzymy model regresji liniowej
    model1 = LinearRegression()
    # Dopasowujemy model do danych treningowych
    model1.fit(X_train_subset, y_train)
    # Przewidujemy oceny 'Toy Story' dla zbioru testowego
    y_pred1 = model1.predict(X_test_subset)
    # Zapisujemy predykcje dla wartości m
    predictions[m] = y_pred1

    
for m, y_pred in predictions.items():
    print("Predykcje dla m =", m, ":")
    # Użyj zip(), aby połączyć prawdziwe oceny (y_test) z predykcjami (y_pred)
    for true_rating, predicted_rating in zip(y_test, y_pred):
        print(f"{true_rating}  |  {predicted_rating}")

In [3]:
import numpy as np

In [4]:
x = np.array([[5,0,2,5], [4,1,2,5]]).T
print(x)

[[5 4]
 [0 1]
 [2 2]
 [5 5]]


Obliczamy normę wektorów kolumnowych

In [5]:
np.linalg.norm(x, axis=0)

array([7.34846923, 6.78232998])

Obliczamy znormalizowaną macierz

In [6]:
x/np.linalg.norm(x, axis=0)

array([[0.68041382, 0.58976782],
       [0.        , 0.14744196],
       [0.27216553, 0.29488391],
       [0.68041382, 0.73720978]])

Nasza własna ocena filmów Matrix - 4, Star Wars IV - 3 dla wygody zapisujemy jako wektor kolumnowy

In [7]:
y = [[4],[3]]
print(y)

[[4], [3]]


Obliczamy podobieństwo cosinusowe z każdym użytkownikiem (skorzystamy z mnożenia macierzowego)

In [8]:
z=np.dot(x/np.linalg.norm(x, axis=0), np.array(y)/np.linalg.norm(y))
print(z)

[[0.89819175]
 [0.08846517]
 [0.39466277]
 [0.98665692]]


Normalizujemy otrzymany wektor (będzie on reprezentował nasz profil filmowy)

In [9]:
z/np.linalg.norm(z)

array([[0.64422929],
       [0.06345177],
       [0.28307243],
       [0.70768107]])

Teraz musimy obliczyć podobieństwo cosinusowe między naszym profilem a każdą kolumną macierzy, aby znaleźć takie filmy, które są podobne do naszego profilu. Sortujemy po otrzymanym podobieństwu i dostajemy rekomendacje. Wykorzystując nasz przykład, otrzymujemy


In [10]:
X = x/np.linalg.norm(x, axis=0)
Z = z/np.linalg.norm(z)
np.dot(X.T, Z)

array([[0.99690104],
       [0.99448407]])

Stąd dostaliśmy większą rekomendacje dla filmu Matrix (0.996) niż Star Wars IV (0.994). Co było oczywiście oczekiwane
skoro nasza ocena była Matrix - 4, a Star Wars IV - 3.

Możemy, teraz przejść do właściwej części zadania.


### Zadanie do wykoniania

Napisz system rekomendacji filmów który będzie wykorzystywał dane
[MovieLens Latest Datasets](https://grouplens.org/datasets/movielens/latest/).
Dokładnie mniejszy zbiór (który dodatkowo trochę jeszcze zmniejszymy). Pobierz plik
[ml-latest-small.zip](http://files.grouplens.org/datasets/movielens/ml-latest-small.zip).
Interesować, będą nas głównie dwa pliki movies.csv oraz ratings.csv. W pliku ratings.csv mamy
właściwie wszystkie dane, które będą nam potrzebne aby stworzyć macierz oceny. Ponieważ
nawet w tym przypadku otrzymana macierz będzie dość duża dlatego w trakcie wczytywania
danych z pliku proszę wziąć pod uwagę tylko te wiersze (z pliku ratings.csv) w których <b>movie_id</b> (druga kolumna)
jest mniejsze od 10000 (resztę ignorujemy). Dla aktualnych danych ze strony otrzymamy, wtedy macierz mniej więcej 611x9019.
Przykładowy wynik dla wybranego profilu filmowego:

```
# wektor my_ratings odpowiada wektorowi y z przykładu wyżej
my_ratings = np.zeros((9019,1))
my_ratings[2571] = 5      # patrz movies.csv  2571 - Matrix
my_ratings[32] = 4        # 32 - Twelve Monkeys
my_ratings[260] = 5       # 260 - Star Wars IV
my_ratings[1097] = 4
my_ratings_norm = my_ratings/np.linalg.norm(my_ratings)

...

# otrzymujemy wynik rekomendacji po posortowaniu
# (cos(θ), movies_id)
(0.8675507828468105, 260)
(0.8362098349303669, 2571)
(0.8227451213877744, 1196)
(0.8002349214247857, 1210)
(0.7458504689612442, 1097)
(0.7286029159733108, 32)
(0.7265369898748615, 1198)
(0.7095672455110477, 1240)
(0.7029872178855614, 1270)
```

Otrzymane reprezentacje wyświetl w postaci nazw filmów, korzystając z movies.csv. Uwaga: w rzeczywistych
danych otrzymamy dużą liczbę zer, nawet całe kolumny, wtedy dostaniemy wartości NaN przy dzieleniu!
Rozwiąż ten problem wykorzystując np.nan_to_num(...).



In [11]:
import numpy as np
import pandas as pd

# Wczytanie danych z plików
ratings_df = pd.read_csv("ratings_small.csv")
movies_df = pd.read_csv("movies.csv")

# Wybór tylko tych wierszy, gdzie movieId < 10000
ratings_df = ratings_df[ratings_df['movieId'] < 10000]

# Tworzenie macierzy ocen
ratings_table = ratings_df.pivot(index='userId', columns='movieId', values='rating').fillna(0)

# Tworzenie przekształconej do tabeli macierzy ocen 
ratings_matrix = ratings_df.pivot(index='userId', columns='movieId', values='rating').fillna(0).to_numpy()
#print(ratings_matrix)

my_ratings = np.zeros((ratings_matrix.shape[1], 1))

my_ratings[ratings_table.columns.get_loc(2571)] = 5      # Matrix
my_ratings[ratings_table.columns.get_loc(32)] = 4        # Twelve Monkeys
my_ratings[ratings_table.columns.get_loc(260)] = 5       # Star Wars IV
my_ratings[ratings_table.columns.get_loc(1097)] = 4      # E.T. the Extra-Terrestrial

#print(my_ratings.shape)

# Obliczenie normy wektorów kolumnowych macierzy
column_norms = np.linalg.norm(ratings_matrix, axis=0)

# Stworzenie znormalizowanej macierzy
normalized_matrix = np.nan_to_num(ratings_matrix / column_norms)
print("normalized matrix ",normalized_matrix.shape)

# Normalizacja wektora ocen konkretnego użytkownika
my_ratings_norm = np.nan_to_num(my_ratings / np.linalg.norm(my_ratings))

# Obliczenie podobieństwa kosinusowego między użytkownikami a użytkownikiem o znormalizowanych preferencjach -> wektor
user_similarity_scores = np.dot(normalized_matrix, my_ratings_norm)

# Znormalizowanie wektora podobieństw miedzy użytkownikami -> wektor
normalized_user_similarity_scores = np.nan_to_num(user_similarity_scores / np.linalg.norm(user_similarity_scores))

# Obliczenie iloczynu skalarnego miedzy naszym profilem a kazda kolumna macierzy
recommendations = np.dot(normalized_matrix.T, normalized_user_similarity_scores)
#print("recomendations: ", recommendations.flatten())

"""W  `recommendations` indeks $i$ odpowiada $i$-temu filmowi w ratings_matrix (to nie *movieId* !). Sortujemy tablicę malejąco względem podobieństwa, ale tak by zachować informację o indeksach, stąd `np.argsort()`. Indeksy wykorzystujemy do pozyskania faktycznych *movieId* z ratings_table.
"""

matrix_indices_sorted = np.argsort(recommendations.flatten())[::-1]

recommendations_sorted = recommendations[matrix_indices_sorted].flatten()
movie_indices = ratings_table.columns[matrix_indices_sorted]

# Wyświetlenie rekomendacji z nazwami filmów
for i in range(len(recommendations_sorted)):
    movie_id = movie_indices[i]
    movie_title = movies_df[movies_df['movieId'] == movie_id]['title'].values[0]
    print(f"({recommendations_sorted[i]}, {movie_id}, {movie_title})")




normalized matrix  (610, 5390)
(0.8675507828468108, 260, Star Wars: Episode IV - A New Hope (1977))
(0.8362098349303673, 2571, Matrix, The (1999))
(0.8227451213877748, 1196, Star Wars: Episode V - The Empire Strikes Back (1980))
(0.800234921424786, 1210, Star Wars: Episode VI - Return of the Jedi (1983))
(0.7458504689612445, 1097, E.T. the Extra-Terrestrial (1982))
(0.728602915973311, 32, Twelve Monkeys (a.k.a. 12 Monkeys) (1995))
(0.7265369898748618, 1198, Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) (1981))
(0.7095672455110479, 1240, Terminator, The (1984))
(0.7029872178855616, 1270, Back to the Future (1985))
(0.6981492768797292, 1291, Indiana Jones and the Last Crusade (1989))
(0.6973450359077008, 2028, Saving Private Ryan (1998))
(0.6842728560727939, 589, Terminator 2: Judgment Day (1991))
(0.6787406267628278, 2762, Sixth Sense, The (1999))
(0.6779667543055932, 1580, Men in Black (a.k.a. MIB) (1997))
(0.6728379041160386, 296, Pulp Fiction (1994))
(0.6676